extensions/csv/src/main/java/io/deephaven/csv/reading/CsvReader.java

```diff
@@ -0,0 +1,302 @@
+package io.deephaven.csv.reading;
+
+import io.deephaven.csv.densestorage.DenseStorageReader;
+import io.deephaven.csv.densestorage.DenseStorageWriter;
+import io.deephaven.csv.parsers.ParserBase;
+import io.deephaven.csv.sinks.Sink;
+import io.deephaven.csv.parsers.context.SentinelConfiguration;
+import io.deephaven.csv.parsers.Parsers;
+import io.deephaven.csv.sinks.SinkFactory;
+import io.deephaven.csv.containers.CharSlice;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.util.*;
+import java.util.concurrent.*;
+
+public class CsvReader {
+    private final Reader reader;
+    private final SinkFactory sinkFactory;
+    /**
+     * Whether to trim leading and trailing blanks from non-quoted values.
+     */
+    private boolean ignoreSurroundingSpaces = false;
+    /**
+     * Whether to trim leading and trailing blanks from inside quoted values.
+     */
+    private boolean trim = false;
+    /**
+     * Whether the incoming data has column headers.
+     */
+    private boolean hasHeaders = true;
+    /**
+     * The quote character (used when you want field or line delimiters to be interpreted as literal text.
+     */
+    private char quoteChar = '#';
+    /**
+     * The field delimiter (the character that separates one column from the nextd0.
+     */
+    private char fieldDelimiter = ',';
+    /**
+     * Whether to run asynchronously. In particular, the operation of reading the raw file, breaking it into columns,
+     * and storing that column text in memory can run in parallel with parsing the data for a given column, and all the
+     * column parsers can themselves run in parallel.
+     */
+    private boolean async = false;
+    /**
+     * The user-defined set of parsers that participate in type inference. Defaults to ZamboniParsers.DEFAULT
+     */
+    private Set<ParserBase<?>> inference = new HashSet<>(Parsers.DEFAULT);
+    /**
+     * Client-specified headers that can be used to override the existing headers in the input (if hasHeaders is true),
+     * or to provide absent headers (if hasHeaders is false).
+     */
+    private List<String> clientSpecifiedHeaders = new ArrayList<>();
+    /**
+     * Used to force a specific parser for a specific column.
+     */
+    private Map<String, ParserBase<?>> parsers = new HashMap<>();
+    /**
+     * The parser
+     */
+    private ParserBase<?> nullParser;
+    /**
+     * The NullContext holding various null sentinels.
+     */
+    private final SentinelConfiguration nullContext = new SentinelConfiguration();
+
+    public CsvReader(Reader reader, SinkFactory sinkFactory) {
+        this.reader = reader;
+        this.sinkFactory = sinkFactory;
+    }
+
+    public Result read() throws IOException {
+        final CellGrabber grabber = new CellGrabber(reader, quoteChar, fieldDelimiter, ignoreSurroundingSpaces, trim);
+        String[] headersToUse = null;
```

Comment:
I realized after reading some more that this will be too complicated, so nevermind.  Leaving the text below for context for you, but please just resolve this comment after you read it.

It may be possible to make this final adding an else to the if below, and capturing the logic of the following if in a slightly different way in the if body.   Just asking you to take a look if that ends up with code that is harder to read; if not, I think is worth it.  If you think the resulting code is worse, feel free to drop this suggestion.

Reply:
I'm able to do it without much pain for firstDataRow. It's harder to do for headersToUse without refactoring the code or making a new method. Eh I think I'll leave it as is.